src/services/TestResultsService.ts

```diff
@@ -860,10 +860,14 @@ export class TestResultsService {
     if (payload.testStatus === TEST_STATUS.SUBMITTED) {
       payload.testTypes.forEach((testType) => {
         // CVSB-7675 if vehicle type is HGV/TRL and testTypeId is Roadworthiness test and testResult is pass then testNumber = certificateNumber
-       if (TestResultsService.isPassedRoadworthinessTestForHgvTrl(payload.vehicleType, testType.testTypeId, testType.testResult)
-                    ||
-            (TestResultsService.isAnnualTestTypeClassificationWithoutAbandonedResult(testType.testTypeClassification, testType.testResult) && !this.isTestTypeAdr(testType) && !this.isTestTypeLec(testType))
-                  ) {
+        // CVSB-13638 Roadworthiness FAIL should not generate a certificate number
+        const isTestTypeRoadworthiness = TestResultsService.isHGVTRLRoadworthinessTest(testType.testTypeId);
+        if (TestResultsService.isPassedRoadworthinessTestForHgvTrl(payload.vehicleType, testType.testTypeId, testType.testResult)
+          ||
+          (TestResultsService.isAnnualTestTypeClassificationWithoutAbandonedResult(testType.testTypeClassification, testType.testResult)
+            && !this.isTestTypeAdr(testType)
+            && !this.isTestTypeLec(testType)
+            && !isTestTypeRoadworthiness)) {
```

Comment:
This if statement gives me a headache.  Rewrite it so that the logic is clearer.

The existing logic says it caters for checking the roadworthiness test is for HGV/Trailer and also a pass.  Is that logic wrong?

Reply:
The existing logic was right. The problem is that even if the `isPassedRoadworthinessTestForHgvTrl` method was returning false (as expected), the second part of the condition was returning true (which was causing the bug); that's why Cat added the additional `!isTestTypeRoadworthiness` check.
I rewrote the logic. Please tell me if it's clearer now.